File: 백준/Silver/1158. 요세푸스 문제/요세푸스 문제.py
class Node:
    
    def __init__(self,key):
        self.key = key 
        self.next = None
        self.prev = None
        
N,k = map(int,input().split())

if N < 2:
    print('<1>')
else:

    humans = [Node(i) for i in range(N)]

    # 첫 번째 사람과 마지막 사람을 제외하고 앞 사람과 뒷 사람 연결시켜주기
    for i in range(1, N-1):
        humans[i].next = humans[i + 1]
        humans[i].prev = humans[i - 1]

    # 첫 번째 사람과 마지막 사람 연결시켜주기 

    humans[0].next = humans[1]
    humans[0].prev = humans[-1]
    humans[-1].next = humans[0]
    humans[-1].prev = humans[-2]   

    result = []
    target = humans[0] 

    while len(result) < N:
        
        for _ in range(k-1): # 2번의 이동을 거쳐야 함
            target = target.next
        result.append(str(target.key + 1)) # 죽을 사람 죽여 ~~ 0번째부터 사람부터 시작하니 + 1
        target.prev.next = target.next 
        target.next.prev = target.prev 
        
        target = target.next # 죽었으니 그 사람 빼고 시작하자고 
        
    print('<' + ', '.join(result)+'>')

File: 백준/Silver/1158. 요세푸스 문제/요세푸스 문제.js
const fs = require('fs');
const filePath =
  process.platform === 'linux' ? '/dev/stdin' : __dirname + '/input.txt';

const [N, K] = fs.readFileSync(filePath).toString().split(' ').map(Number);

function solution() {
  if (N === 1) {
    console.log('<1>');
    return;
  }

  const arr = Array.from({ length: N }, (_, idx) => {
    return {
      value: idx + 1,
      next: null,
      prev: null,
    };
  });

  /* 양방향 링크드 리스트 형태로 연결 */
  for (let idx = 1; idx <= N - 2; idx += 1) {
    arr[idx].next = arr[idx + 1];
    arr[idx].prev = arr[idx - 1];
  }
  arr[N - 1].next = arr[0];
  arr[N - 1].prev = arr[N - 2];

  arr[0].next = arr[1];
  arr[0].prev = arr[N - 1];

  const result = [];
  let cnt = 0;

  /* K 번째 사람 이미 사망 처리 하기 */

  let target = arr[K - 1];
  target.prev.next = target.next;
  target.next.prev = target.prev;

  result.push(target.value);
  cnt += 1;

  while (cnt < N) {
    for (let step = 0; step < K; step += 1) {
      target = target.next;
    }

    target.prev.next = target.next;
    target.next.prev = target.prev;
    result.push(target.value);
    cnt += 1;
  }

  console.log('<' + result.join(', ') + '>');
}

solution();
